libs/vectorise/include/vectorise/uint/int.hpp

```diff
@@ -0,0 +1,1298 @@
+ *
+ * The implementation subclasses a <byte_array::ConstByteArray> such
+ * one easily use this in combination with hashes etc.
+ */
+template <uint16_t S = 256>
+class Int
+{
+public:
+  using BaseType = uint8_t;
+  using WideType = uint64_t;
+  enum
+  {
+    INT_SIZE          = S,
+    ELEMENT_SIZE      = sizeof(BaseType) * 8,
+    ELEMENTS          = INT_SIZE / ELEMENT_SIZE,
+    WIDE_ELEMENT_SIZE = sizeof(WideType) * 8,
+    WIDE_ELEMENTS     = (INT_SIZE + WIDE_ELEMENT_SIZE - 1) / WIDE_ELEMENT_SIZE,
+  };
+  static_assert(
+      S >= 256,
+      "Int class is intended for ints >= 256 bits, for smaller sizes, use the primitive types");
+  static_assert(S == (WIDE_ELEMENTS * WIDE_ELEMENT_SIZE), "Size must be a multiple of 64 bits.");
+
+  using WideContainerType                   = core::Array<WideType, WIDE_ELEMENTS>;
+  using ContainerType                       = BaseType[ELEMENTS];
+  static constexpr char const *LOGGING_NAME = "Int";
+
+  template <typename... T>
+  using IfIsWideInitialiserList = std::enable_if_t<meta::Is<WideType>::SameAsEvery<T...>::value &&
+                                                   (sizeof...(T) <= WIDE_ELEMENTS)>;
+
+  template <typename... T>
+  using IfIsBaseInitialiserList =
+      std::enable_if_t<meta::Is<BaseType>::SameAsEvery<T...>::value && (sizeof...(T) <= ELEMENTS)>;
+
+  ////////////////////
+  /// constructors ///
+  ////////////////////
+
+  constexpr Int()                     = default;
+  constexpr Int(Int const &other)     = default;
+  constexpr Int(Int &&other) noexcept = default;
+
+  template <typename... T, IfIsWideInitialiserList<T...> * = nullptr>
+  constexpr explicit Int(T &&... data)
+    : wide_{{std::forward<T>(data)...}}
+  {}
+
+  template <typename... T, IfIsBaseInitialiserList<T...> * = nullptr>
+  constexpr explicit Int(T &&... data)
+    : wide_{reinterpret_cast<WideContainerType &&>(
+          core::Array<BaseType, ELEMENTS>{{std::forward<T>(data)...}})}
+  {}
+
+  template <typename T, meta::IfIsAByteArray<T> * = nullptr>
+  explicit Int(T const &other);
+  template <typename T, meta::IfIsSignedInteger<T> * = nullptr>
+  constexpr explicit Int(T number);
+  template <typename T, meta::IfIsUnsignedInteger<T> * = nullptr>
+  constexpr explicit Int(T number);
+  constexpr explicit Int(__int128_t number);
+  constexpr explicit Int(__uint128_t number);
+
+  /////////////////////////
+  /// casting operators ///
+  /////////////////////////
+
+  explicit operator std::string() const;
+  template <typename T, meta::IfIsInteger<T> * = nullptr>
+  explicit constexpr operator T() const;
+  explicit constexpr operator __int128_t() const;
+  explicit constexpr operator __uint128_t() const;
+
+  ////////////////////////////
+  /// assignment operators ///
+  ////////////////////////////
+
+  constexpr Int &operator=(Int const &v);
+  template <typename ArrayType>
+  meta::IfHasIndex<ArrayType, Int> &operator=(ArrayType const &v);  // NOLINT
```

Comment:
what problem does the NOLINT solve here?

Reply:
it's a false alarm: 

> Line: 123 @ 3
>   Message: operator=() should return 'Int&'
>   Check: misc-unconventional-assign-operator
>   Context:
>       meta::IfHasIndex<ArrayType, Int> &operator=(ArrayType const &v);